lib/elixir/lib/calendar/datetime.ex

```diff
@@ -240,6 +439,170 @@ defmodule DateTime do
+      {:ok, true} ->
+        :ok
+
+      {:ok, false} ->
+        {:error, :invalid_leap_second}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  @doc """
+  Changes the time zone of a `DateTime`.
+
+  Returns a `DateTime` for the same point in time, but instead at the time zone
+  provided.
+
+  Requires passing a `TimeZoneDatabase` as an argument or setting it with
+  `TimeZoneDatabaseClient.set_database/1`.
+
+  ## Examples
+
+      iex> cph_datetime = DateTime.from_naive!(~N[2018-07-16 12:00:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, pacific_datetime} = DateTime.shift_zone(cph_datetime, "America/Los_Angeles", FakeTimeZoneDatabase)
+      iex> pacific_datetime
+      #DateTime<2018-07-16 03:00:00-07:00 PDT America/Los_Angeles>
+
+  """
+  @doc since: "1.8.0-dev"
+  @spec shift_zone(t, Calendar.time_zone(), TimeZoneDatabaseClient.tz_db_or_config()) ::
+          {:ok, t} | {:error, :time_zone_not_found} | {:error, atom}
+  def shift_zone(datetime, time_zone, tz_db_or_config \\ :from_config)
+
+  def shift_zone(%{time_zone: time_zone} = datetime, time_zone, _) do
+    # When the desired time_zone is the same as the existing time_zone just return it unchanged.
+    {:ok, datetime}
+  end
+
+  def shift_zone(%{second: 60, calendar: Calendar.ISO} = datetime, time_zone, tz_db_or_config) do
+    # If second is 60 (positive leap second) adjust it to 59, calculate, then adjust back to 60.
+    case shift_zone(%{datetime | second: 59}, time_zone, tz_db_or_config) do
+      {:ok, %{second: second} = dt_result} when second == 59 ->
+        {:ok, %{dt_result | second: 60}}
+
+      {:ok, _} ->
+        {:error, :non_whole_minute_offsets_not_supported_for_leap_seconds}
+
+      error ->
+        error
+    end
+  end
+
+  def shift_zone(
+        %{
+          calendar: Calendar.ISO,
+          std_offset: std_offset,
+          utc_offset: utc_offset,
+          microsecond: {_, microsecond_precision}
+        } = datetime,
+        time_zone,
+        tz_db_or_config
+      ) do
+    datetime_in_utc_iso_days =
+      datetime
+      |> to_iso_days()
+      |> apply_tz_offset(utc_offset + std_offset)
+
+    case TimeZoneDatabaseClient.time_zone_period_from_utc_iso_days(
+           datetime_in_utc_iso_days,
+           time_zone,
+           tz_db_or_config
+         ) do
+      {:ok, period} ->
+        naive_datetime =
+          datetime_in_utc_iso_days
+          |> apply_tz_offset(-period.utc_offset - period.std_offset)
+          |> iso_days_to_iso_naive_datetime(microsecond_precision)
+
+        do_from_naive(
+          naive_datetime,
+          time_zone,
+          period.std_offset,
+          period.utc_offset,
+          period.zone_abbr
+        )
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  def shift_zone(%{calendar: calendar} = datetime, time_zone, tz_db_or_config)
+      when calendar != Calendar.ISO do
```

Comment:
Since shift_zone relies on iso days conversion, we can completely skip those conversions and generalize the clause above. I have a patch ready.